### PR TITLE
Support for project branches

### PR DESCRIPTION
**Pull request by DatzAtWork**
# Proposed changes
* Implementation for feature request #96 

This pull request adds support for the **branches feature** in the commercial sonarqube editions and the [sonarqube-community-branch-plugin](https://github.com/mc1arke/sonarqube-community-branch-plugin).
It adds the (optional) new command line parameter **-b** which can be used to select a specific branch (see README.md).
If the parameter is not set, the default branch is exported (like before).

**Note**: The plugin web-form is not extended with this pull request. The new feature can only be used from the command line or rest-service.

**Travis**: I had to adjust the `.travis.yml` to make the build work with java8.

### CODE REVIEW

**Comment by DatzAtWork** [comment]
Travis-CD fails for `sonarqube=none`. I am not sure how you made it run. :-/

**Comment by Sancretor** [comment]
Hi @DatzAtWork 

Thanks a lot for your contribution, we really appreciate it !
I also have to apologize for the long delay... we had a hard time finding time to manage this project.

Regarding the Travis-CI failure, it's a know problem in our configuration to analyze code from PR with SonarCloud.
No need for you to waste time with that...

File: README.md [hunk 1]
```diff
@@ -32,11 +32,12 @@ This tool can be used in standalone as a JAR executable (with the command line)
 #### Get help
 Use `java -jar cnesreport.jar -h` to get the following help about cnesreport:
 ````
-usage: java -jar cnesreport.jar [-a <arg>] [-c] [-d <arg>] [-e] [-h] [-l <arg>] [-o <arg>] [-p <arg>] [-r <arg>]
+usage: java -jar cnesreport.jar [-a <arg>] [-b <arg>] [-c] [-d <arg>] [-e] [-h] [-l <arg>] [-o <arg>] [-p <arg>] [-r <arg>]
        [-s <arg>] [-t <arg>] [-v] [-w] [-x <arg>]
 Generate editable reports for SonarQube projects.
 
  -a,--author <arg>                 Name of the report writer.
+ -b,--branch <arg>                 Branch of the targeted project. Requires Developer Edition or sonarqube-community-branch-plugin.
```
Comment (Sancretor) [inline README.md]: Could you please write what the default value is ?
Just like the "date" parameter for example.

File: README.md [hunk 2]
```diff
@@ -77,6 +78,12 @@ If you are using a secured instance of SonarQube, you can provide a SonarQube au
 java -jar cnesreport.jar -t xuixg5hub345xbefu -s https://example.org:9000 -p projectId -r ./template.docx
 ````
 
+##### Export of a specific project branch (standalone)
+If you are using a commercial edition of sonarqube or the [sonarqube-community-branch-plugin](https://github.com/mc1arke/sonarqube-community-branch-plugin) you want to export the report for a specific branch of your project using the -b option.
```
Comment (Sancretor) [inline README.md]: Could you please replace " **you want to** export the report" by "**, you can** export the report" ?
Just in order to match the sentence format of the other sections.

File: src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java [hunk 3]
```diff
@@ -47,6 +47,7 @@
             {"s", "server", Boolean.TRUE.toString(), "Complete URL of the targeted SonarQube server."},
             {"t", "token", Boolean.TRUE.toString(), "SonarQube token of the SonarQube user who has permissions on the project."},
             {"p", "project", Boolean.TRUE.toString(), "SonarQube key of the targeted project."},
+            {"b", "branch", Boolean.TRUE.toString(), "Branch of the targeted project. Requires Developer Edition or sonarqube-community-branch-plugin."},
```
Comment (Sancretor) [inline src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java]: Again, could you please specify the default value in the description ?

File: src/main/java/fr/cnes/sonar/report/utils/StringManager.java [hunk 4]
```diff
@@ -38,6 +38,8 @@
     public static final String TAB = "\t";
     /** Just a space. */
     public static final String SPACE = " ";
+    /** Placeholder for no-branch. */
+    public static final String NO_BRANCH = "%";
```
Comment (Sancretor) [inline src/main/java/fr/cnes/sonar/report/utils/StringManager.java]: I wonder if setting the default value to "**master**" would be better here, for the sake of clarity.
@begarco & @DatzAtWork : What do you think about it ?

File: .travis.yml [hunk 5]
```diff
@@ -1,5 +1,5 @@
+dist: xenial
```
Comment (Sancretor) [inline .travis.yml]: Could you please revert your modifications on the `.travis.yml` file ?
What you did was already merged on `lequal:dev`.

**Comment by Sancretor** [comment]
As this feature should also be available for the plugin version, I created a dedicated issue : #112

File: README.md (hunk 1 above)
Comment (DatzAtWork) [reply README.md]: Default is nothing, which means *do not use branches*. How do you think, we should call this default value? `(defaults to using no branches)`?

File: README.md (hunk 2 above)
Comment (DatzAtWork) [reply README.md]:
```suggestion
If you are using a commercial edition of sonarqube or the [sonarqube-community-branch-plugin](https://github.com/mc1arke/sonarqube-community-branch-plugin) you can export the report for a specific branch of your project using the -b option. 
```

File: src/main/java/fr/cnes/sonar/report/utils/StringManager.java (hunk 4 above)
Comment (DatzAtWork) [reply src/main/java/fr/cnes/sonar/report/utils/StringManager.java]: I guess this won't work if the branches are not supported by the Sonarqube installation.
No branches is not the same as master-branch.

File: .travis.yml (hunk 5 above)
Comment (DatzAtWork) [reply .travis.yml]: (I will, but this will break all builds.)

Comment (Sancretor) [reply .travis.yml]: I know... but this has already been done on the dev branch.
We won't bother you with this broken build, as you're not responsible of this issue.

File: README.md (hunk 1 above)
Comment (Sancretor) [reply README.md]: I asked because I thought that maybe changing the default to master would be useful.
It may be useless in the end, I agree.

Comment (DatzAtWork) [reply README.md]: Behaviour:
**SonarQube with branches-support**
`-b` is set --> the branch of the project (`-p`) is used
`-b` is not set --> the *main branch* of the project (`-p`) is used (mostly `master` but can be changed in SQ)
**SonarQube without branches-support**
`-b` is set --> undefined behaviour
`-b` is not set --> the project (`-p`) is used

File: src/main/java/fr/cnes/sonar/report/utils/StringManager.java (hunk 4 above)
Comment (Sancretor) [reply src/main/java/fr/cnes/sonar/report/utils/StringManager.java]: I have to test again it to be sure, but I guess that without the branch feature, the api just ignores the branch parameter.
My concern is also that, with the branch feature, if you do not provide the branch parameter, it won't work either..

File: README.md (hunk 1 above)
Comment (DatzAtWork) [reply README.md]:
```suggestion
 -b,--branch <arg>                 Branch of the targeted project. Requires Developer Edition or sonarqube-community-branch-plugin. Default: usage of main branch.
```

File: src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java (hunk 3 above)
Comment (DatzAtWork) [reply src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java]:
```suggestion
            {"b", "branch", Boolean.TRUE.toString(), "Branch of the targeted project. Requires Developer Edition or sonarqube-community-branch-plugin. Default: usage of main branch."},
```

File: src/main/java/fr/cnes/sonar/report/utils/StringManager.java (hunk 4 above)
Comment (Sancretor) [reply src/main/java/fr/cnes/sonar/report/utils/StringManager.java]: I just did some tests with the *community branch plugin*, and you're right.
With no branch parameter, or with your default value `%` , it returns data on the *main branch*.
Indeed, setting the default to `master` would be an error in the case our main branch is named differently...
Your work seems OK to me !

File: src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java (hunk 3 above)
Comment (Sancretor) [reply src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java]: Thanks, looks good to me.

File: README.md (hunk 1 above)
Comment (Sancretor) [reply README.md]: Thanks, looks good to me.

Comment (Sancretor) [reply README.md]: Could you please commit the suggested changes ?  
Then I will validate the pull request.

File: src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java (hunk 3 above)
Comment (Sancretor) [reply src/main/java/fr/cnes/sonar/report/utils/CommandLineManager.java]: Could you please commit the suggested changes ?
Then I will validate the pull request.

File: src/main/java/fr/cnes/sonar/report/utils/StringManager.java (hunk 4 above)
Comment (Sancretor) [reply src/main/java/fr/cnes/sonar/report/utils/StringManager.java]: Could you please commit the suggested changes ?
Then I will validate the pull request.

**Pull request merged**